tools/xdp/s2n-quic-xdp/src/if_xdp.rs

```diff
@@ -0,0 +1,265 @@
+#[repr(C)]
+pub struct RingOffsetV1 {
+    pub producer: u64,
+    pub consumer: u64,
+    pub desc: u64,
+}
+
+/// A structure to communicate the offsets for an individual ring for an AF_XDP socket
+///
+/// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L43
+#[derive(Clone, Copy, Debug, Default)]
+#[repr(C)]
+pub struct RingOffsetV2 {
+    pub producer: u64,
+    pub consumer: u64,
+    pub desc: u64,
+    pub flags: u64,
+}
+
+impl RingOffsetV2 {
+    /// Converts an offset description from an older version
+    #[inline]
+    fn set_v1(&mut self, v1: RingOffsetV1) {
+        // Logic from https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L197
+        self.producer = v1.producer;
+        self.consumer = v1.consumer;
+        self.desc = v1.desc;
+        self.flags = v1.consumer + (size_of::<u32>() as u64);
+    }
+}
+
+/// A structure to communicate the offsets of each of the 4 rings for an AF_XDP socket
+///
+/// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L50
+#[derive(Clone, Copy, Debug, Default)]
+#[repr(C)]
+pub struct MmapOffsets<O = RingOffsetV2> {
+    pub rx: O,
+    pub tx: O,
+    pub fill: O,
+    pub completion: O,
+}
+
+impl MmapOffsets {
+    /// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L92
+    pub const RX_RING: usize = 0;
+    /// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L93
+    pub const TX_RING: usize = 0x8000_0000;
+    /// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L94
+    pub const FILL_RING: usize = 0x1_0000_0000;
+    /// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L95
+    pub const COMPLETION_RING: usize = 0x1_8000_0000;
+
+    #[inline]
+    pub(crate) fn as_v1(mut self) -> Self {
+        // getsockopt on a kernel <= 5.3 has no flags fields.
+        // Copy over the offsets to the correct places in the >=5.4 format
+        // and put the flags where they would have been on that kernel.
+
+        // Logic from https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L197
+        let v1 = unsafe { *(&self as *const Self as *const MmapOffsets<RingOffsetV1>) };
+        self.rx.set_v1(v1.rx);
+        self.tx.set_v1(v1.tx);
+        self.fill.set_v1(v1.fill);
+        self.completion.set_v1(v1.completion);
+
+        self
+    }
+}
+
+/// Socket option ids passed to an AF_XDP socket with the `setsockopt` syscall.
+///
+/// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L58-L65
+#[derive(Clone, Copy, Debug)]
+#[repr(i32)]
```

Comment:
Can you explain when you would use `repr[u/i]`? For example what would be the difference between a `i32` or `u32` here?

Reply:
These flags are passed into the `setsockopt` `optname` parameter, which is an `int` (see https://linux.die.net/man/2/setsockopt). I can add a comment saying that.

Reply:
Since most of the others in this file are repr C, it might be useful. Especially since we are interfacing with C and there is alot of magic.